build-system/tasks/runtime-test/helpers.js

```diff
@@ -249,9 +289,402 @@ function isLargeRefactor() {
+      ' changed in the local branch.',
+  };
+  if (argv.chrome_flags) {
+    log(
+      green('Launching'),
+      cyan(CHROMEBASE),
+      green('with flags'),
+      cyan(chromeFlags)
+    );
+  }
+  if (!isTravisBuild()) {
+    log(
+      green('Run'),
+      cyan('gulp help'),
+      green('to see a list of all test flags.')
+    );
+    log(green('⤷ Use'), cyan('--nohelp'), green('to silence these messages.'));
+    if (
+      !argv.unit &&
+      !argv.integration &&
+      !argv.files &&
+      !argv.a4a &&
+      !argv['local-changes']
+    ) {
+      log(green('Running all tests.'));
+      log(
+        green('⤷ Use'),
+        cyan('--unit'),
+        green('or'),
+        cyan('--integration'),
+        green('to run just the unit tests or integration tests.')
+      );
+      log(
+        green('⤷ Use'),
+        cyan('--local-changes'),
+        green('to run unit tests from files commited to the local branch.')
+      );
+    }
+    if (!argv.testnames && !argv.files && !argv.local_changes) {
+      log(
+        green('⤷ Use'),
+        cyan('--testnames'),
+        green('to see the names of all tests being run.')
+      );
+    }
+    if (!argv.headless) {
+      log(
+        green('⤷ Use'),
+        cyan('--headless'),
+        green('to run tests in a headless Chrome window.')
+      );
+    }
+    if (!argv.compiled) {
+      log(green('Running tests against unminified code.'));
+    }
+    Object.keys(argv).forEach(arg => {
+      const message = argvMessages[arg];
+      if (message) {
+        log(yellow(`--${arg}:`), green(message));
+      }
+    });
+  }
+}
+
+function maybePrintCoverageMessage() {
+  if (!argv.coverage || isTravisBuild()) {
+    return;
+  }
+
+  const url = 'file://' + path.resolve('test/coverage/index.html');
+  log(green('INFO:'), 'Generated code coverage report at', cyan(url));
+  opn(url, {wait: false});
+}
+
+function maybeSetCoverageConfig(config, reportName) {
+  if (!argv.coverage) {
+    return;
+  }
+
+  config.plugins.push('karma-coverage-istanbul-reporter');
+  config.coverageIstanbulReporter = {
+    dir: 'test/coverage',
+    reports: isTravisBuild() ? ['lcovonly'] : ['html', 'text', 'text-summary'],
+    'report-config': {lcovonly: {file: reportName}},
+  };
+
+  config.browserify.transform = [
+    [
+      'babelify',
+      {
+        plugins: [
+          [
+            'istanbul',
+            {
+              exclude: [
+                'ads/**/*.js',
+                'third_party/**/*.js',
+                'test/**/*.js',
+                'extensions/**/test/**/*.js',
+                'testing/**/*.js',
+              ],
+            },
+          ],
+        ],
+      },
+    ],
+  ];
+}
+
+function karmaBrowserComplete(browser) {
+  const result = browser.lastResult;
+  // Prevent cases where Karma detects zero tests and still passes. #16851.
+  if (result.total == 0) {
+    log(red('ERROR: Zero tests detected by Karma.'));
+    log(red(JSON.stringify(result)));
```

Comment:
Note for a later bug fix: It turns out that this occasionally happens even though tests are actually running, and a corrected result is forthcoming. (Not sure how to fix it, since there are other times when this indicates a total Karma / Sauce failure.)